lib/ansible/galaxy/dependency_resolution/gpg.py

```diff
@@ -0,0 +1,298 @@
+# -*- coding: utf-8 -*-
+# Copyright: (c) 2022, Ansible Project
+# GNU General Public License v3.0+ (see COPYING or https://www.gnu.org/licenses/gpl-3.0.txt)
+"""Signature verification helpers."""
+
+from __future__ import (absolute_import, division, print_function)
+__metaclass__ = type
+
+from ansible.errors import AnsibleError
+from ansible.module_utils.common.text.converters import to_text
+from ansible.module_utils.urls import open_url
+from ansible.utils.display import Display
+
+import os
+import subprocess
+import textwrap
+
+from dataclasses import dataclass, fields as dc_fields
+from urllib.error import HTTPError
+
+display = Display()
+
+
+def run_gpg_verify(b_manifest_file, signature, keyring):
+    status_fd_read, status_fd_write = os.pipe()
+
+    cmd = [
+        'gpg',
+        f'--status-fd={status_fd_write}',
+        '--verify',
+        '--batch',
+        '--no-tty',
+        '--no-default-keyring',
+        f'--keyring={keyring}',
+        '-',
+        b_manifest_file,
+    ]
+
+    try:
+        p = subprocess.Popen(
+            cmd,
+            stdin=subprocess.PIPE,
+            stdout=subprocess.PIPE,
+            stderr=subprocess.PIPE,
+            pass_fds=(status_fd_write,),
+        )
+        stdout, stderr = p.communicate(input=signature)
+    except subprocess.SubprocessError as err:
+        cmd_str = ' '.join([to_text(arg) for arg in cmd])
+        display.error(f"Failed during GnuPG verification with command '{cmd_str}':\n{err}")
+        p = None
```

Comment:
Why not just
```suggestion
        return
```

Reply:
I feel like returning in the except clause isn't super readable if there's a `finally`. This should have returned a tuple though. I refactored a bit and I'm raising an AnsibleError instead to halt execution - I don't know why I was using `display.error` to continue regardless of exceptions.

Reply:
Interesting, I've never heard that `except->return` + `finally` situation to be mentioned in the context of readability. I've always felt like this pattern is natural.
I think what's unnatural is that there's a disconnected `if`-block that is coupled to some code in the `except` block 5 lines above. Having a `finally` usually does not relate to how the control flow gets to the `return` but rather it relates to the cleanup of something in the `try` block and so it doesn't really matter when we read "alright, this is getting interrupted because of some foreseen failure". From the readability standpoint, we usually don't need to know about the cleanup part of the flow, unless we're specifically looking for it.

Reply:
Realized I didn't respond to this one.

except->return + finally->return could be a gotcha (I did have a conditional return in the finally at one point).
I find it pretty annoying to skim (or debug) a function with a lot of return statements (it's easy to miss a sneaky return), so in general, having fewer and keeping them together is what I prefer to read. Maybe this is personal-preference, but I find it amazing if you've never tripped over something similar.

This was a technically excessive variable, because I *wanted* to have the return statements towards the bottom of the function. Since the variable was created outside of the try-except, using it following the try-except doesn't seem unnatural to me.

Anyway, this code needed to be tightened up in other ways, so there is no more `p` variable and no more 'disconnected' if block.